Cut the client repo, staging it in the main repo

caesarxuchao opened the pull request:
Tracking issue: #28559
ref: https://github.com/kubernetes/kubernetes/pull/25978#issuecomment-232710174

This PR implements the plan a few of us came up with last week for cutting client into its own repo:
1. creating "staging" (name is tentative) directory in the main repo, using a script to copy the client and its dependencies to this directory
2. periodically publishing the contents of this staging client to k8s.io/client-go repo
3. converting k8s components in the main repo to use the staged client. They should import the staged client as if the client were vendored. (i.e., the import line should be `import "k8s.io/client-go/<pacakge name>`). This requirement is to ease step 4.
4. In the future, removing the staging area, and vendoring the real client-go repo.

The advantage of having the staging area is that we can continuously run integration/e2e tests with the latest client repo and the latest main repo, without waiting for the client repo to be vendored back into the main repo. This staging area will exist until our test matrix is vendoring both the client and the server.

Some **_tricks**_ I did in this PR:  
1. I created a symlink under ./vendor, pointing to the staging area, so packages in the main repo can refer to the client repo as if it's vendored. 
2. To prevent the godep tool from messing up the staging area, I exported the staged client to GOPATH in hack/godep-save.sh, so godep will think the all packages referred by the staging client are local and won't attempt to vendor them.
3. I rearranged the directory layout of clients in a more meaningful way.
4. I converted the the `record` tool (tools/record) to v1 in the copy.sh.

@thockin @lavalamp @bgrant0607 @kubernetes/sig-api-machinery

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/29147)

<!-- Reviewable:end -->

caesarxuchao commented on _staging/src/k8s.io/client-go/copy.sh:
```diff
@@ -0,0 +1,90 @@
+echo "rewriting imports"
+grep -Rl "\"${MAIN_REPO_FROM_SRC}" ./ | grep ".go" | grep -v "vendor/" | xargs sed -i "s|\"${MAIN_REPO_FROM_SRC}|\"${CLIENT_REPO_FROM_SRC}|g"
+
+echo "converting pkg/client/record to v1"
```
This block is automating #24410. It's not required for now.

smarterclayton commented:
I'm not in love with `client-go` but it's the simplest thing that covers the use case.  The util change looked ok.

smarterclayton commented:
I'm also not in love with `k8s.io/client-go/pkg/...` but again, can live with it.

lavalamp commented on test/e2e/framework/framework.go:
```diff
@@ -18,20 +18,25 @@ package framework
 	"sync"
 	"time"
 
+	"k8s.io/client-go/pkg/client/clientset_generated/release_1_4"
```
Would be great if we could make this "k8s.io/client-go"

lavalamp commented on test/e2e/framework/framework.go:
```diff
@@ -18,20 +18,25 @@ package framework
 	"time"
 
+	"k8s.io/client-go/pkg/client/clientset_generated/release_1_4"
+	clientreporestclient "k8s.io/client-go/pkg/client/restclient"
```
And if this could be "k8s.io/client-go/restclient" or just "k8s.io/client-go/rest"

lavalamp commented on test/e2e/generated_clientset.go:
```diff
@@ -243,3 +267,45 @@ var _ = framework.KubeDescribe("Generated release_1_3 clientset", func() {
 	})
 })
+
+var _ = framework.KubeDescribe("Generated release_1_4 clientset", func() {
```
change description

lavalamp commented:
I looked at the commits, but probably I need to just pull it locally and play around with it, since the change is too big for github to display effectively.

caesarxuchao commented:
@lavalamp yeah, pulled it locally is helpful. Perhaps you can help verify if `godep save` mess up the client.

caesarxuchao commented:
@smarterclayton we'll rearrange the directory layout. I'll propose a structure and let you guys review soon. This PR is more about proving the staging area works.

caesarxuchao commented:
I'll figure out why the jenkins builds failed. It build system works locally.

caesarxuchao commented:
It's seems the vendor/k8s.io/client-go symlink doesn't work in jenkins.

caesarxuchao commented:
Regarding how to restructure the client repo, I plan to do the following adjustment:
pkg/client/clientset_generated -> clientset
pkg/client/typed/dynamic -> dynamic
pkg/client/typed/discovery -> discovery
pkg/client/cache -> cache
pkg/client/restclient -> rest
And rename pkg/ to dependencies/

For reference, currently pkg/client/ in the client repo looks like this:
pkg/client/
├── clientset_generated
│   └── release_1_4
│      └── ...
├── cache
├── metrics
├── record
├── restclient
├── testing
│   └── core
├── transport
├── typed
│   ├── discovery
│   │   └── fake
│   └── dynamic
└── unversioned
    ├── auth
    ├── clientcmd
    │   └── api
    │       ├── latest
    │       └── v1
    └── portforward

lavalamp commented:
How about:

/pkg/client/clientset_generated/release_1_4 -> /
pkg/client/cache -> tools/cache

and also:
pkg/client/transport -> transport
pkg/client/record -> tools/record
pkg/client/unversioned -> delete

metrics - I need to check what this is but move under tools if it makes sense to keep it?

smarterclayton commented:
Since this is a library I think it's ok to treat it like a library (not use
`pkg/`).  I do think cache could easily be a top level package.

On Tue, Jul 19, 2016 at 8:30 PM, Daniel Smith <notifications@github.com>
wrote:

> How about:
>
> /pkg/client/clientset_generated/release_1_4 -> /
> pkg/client/cache -> tools/cache
>
> and also:
> pkg/client/transport -> transport
> pkg/client/record -> tools/record
> pkg/client/unversioned -> delete
>
> metrics - I need to check what this is but move under tools if it makes
> sense to keep it?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/29147#issuecomment-233806362>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p_QB-yS4twlRNO1jESdqvvXENYSgks5qXWwngaJpZM4JPNRl>
> .
>

caesarxuchao commented:
@lavalamp do we really want to move the pkg/client/unversioned to a top level folder? Here's its layout:
pkg/client/unversioned/
├── auth
├── clientcmd
│   └── api
│       ├── latest
│       └── v1
└── portforward

caesarxuchao commented:
If we move `/pkg/client/clientset_generated/release_1_4` to `/`, we need to find a better name than `client-go`, because package name cannot contain `-`, while the our tradition of repo naming is using `-` to concatenate words.

philips commented:
cc @xiang90 @hongchaodeng might be useful for the etcd TPR

caesarxuchao commented:
One remaining item is finding a proper package name for the clientset. Two possible layouts:
1. move it to the top level, but because the package name cannot contain `-`, we can name the package as `client` while keep the repo name `client-go`
2. or we can move it to ./client, then the directory name is the same as the package name, though the directory name is verbose (`k8s.io/client-go/client`)

caesarxuchao commented:
Top 2 levels of the client repo:
```
├── client
│   ├── fake
│   └── typed
├── discovery
│   └── fake
├── dynamic
├── pkg
│   ├── api
│   ├── apimachinery
│   ├── apis
│   ├── auth
│   ├── capabilities
│   ├── client
│   ├── conversion
│   ├── federation
│   ├── fields
│   ├── kubelet
│   ├── labels
│   ├── master
│   ├── runtime
│   ├── third_party
│   ├── types
│   ├── util
│   ├── version
│   └── watch
├── rest
├── tools
│   ├── auth
│   ├── cache
│   ├── clientcmd
│   ├── portforward
│   └── record
└── transport
```

smarterclayton commented:
I prefer for package names to always match their directory - I hate
go-dockerclient for that reason (it forces me to go look it up).

I'd recommend client.  Or /kclient or /kubeclient

caesarxuchao commented:
@smarterclayton It's under ./client now.

kostya-sh commented:
IMHO `client` is not a great name for the client library package. When reading the code `client.New` doesn't tell anything. Is it kubernetes client? Or etcd client? Or some other client?

Similarly package name `v1` is not very meaningful either. In the code it will always be renamed in the import statement (like here https://github.com/caesarxuchao/kubernetes/blob/86e1285956f6db507a331357a3bbab70814544cf/_staging/src/k8s.io/client-go/client/clientset.go#L21).

Just my 2c.

lavalamp commented:
What about just "kubernetes" for the client directory? or "k8s"?

lavalamp commented:
Can you add a README.md file to the root of the client repo?

lavalamp commented:
There's still a pkg/client/metrics and testing directories. Can we make them go somewhere else?

lavalamp commented on _staging/src/k8s.io/client-go/copy.sh:
```diff
@@ -0,0 +1,146 @@
+echo "rearranging directory layout"
+# $1 and $2 are relative to ${CLIENT_REPO}
+function mvfolder {
+local src=${1%/#/}
```
Indentation in this function :)

caesarxuchao replied:
Done.

lavalamp commented:
```sh
$ cd _staging/src/k8s.io/client-go
$ export SGP=path/to/_staging
$ find | sed 's/^[.][/]\(.*\)[/].*[.]go$/\1/' | grep -v '[.]/' | sort | uniq > existing.txt
$ for i in `GOPATH=$SGP go list -f "{{.Deps}}" k8s.io/client-go/client k8s.io/client-go/tools/...`; do echo $i; done | grep k8s.io/client-go | sed 's|k8s.io/client-go/||' | sort | uniq > imported.txt
```

There's a lot more existing than is imported. Is my Deps call wrong or can we delete some stuff? can you turn the above shell into something that the copy script does to automatically delete all unreferenced packages?

caesarxuchao commented:
> What about just "kubernetes" for the client directory? or "k8s"?

I tend to rename it as k8sclient. We may have other k8s repos in the future so we'd better avoid potential conflict.

> There's still a pkg/client/metrics and testing directories. Can we make them go somewhere else?

Because both packages contain user-facing functions, I moved pkg/client/metrics to tools/metrics and pkg/client/testing/core to testing/.

caesarxuchao commented:
I twisted the bash script a little:
```
cd _staging/src/k8s.io/client-go
export SGP=path/to/_staging
find | sed 's/^[.][/]\(.*\)[/].*[.]go$/\1/' | grep -v '[.]/' | sort | uniq > existing.txt
for i in `GOPATH=$SGP go list -f "{{.Deps}}" k8s.io/client-go/k8sclient k8s.io/client-go/tools/... k8s.io/client-go/tools/cache k8s.io/client-go/discovery k8s.io/client-go/dynamic k8s.io/client-go/rest k8s.io/client-go/transport`; do echo $i; done | grep k8s.io/client-go | sed 's|k8s.io/client-go/||' | sort | uniq > imported.txt
diff --new-line-format="" --unchanged-line-format=""  existing.txt imported.txt | grep -v "tools/\|k8sclient\|testing\|discovery\|dynamic"
```
and the output is
```
pkg/apis/apps/install
pkg/apis/apps/v1alpha1
pkg/apis/certificates/install
pkg/apis/certificates/v1alpha1
pkg/apis/componentconfig
pkg/apis/componentconfig/install
pkg/apis/componentconfig/v1alpha1
pkg/apis/policy
pkg/apis/policy/install
pkg/apis/policy/v1alpha1
pkg/apis/rbac
pkg/apis/rbac/install
pkg/apis/rbac/v1alpha1
pkg/api/testapi
pkg/federation/apis/federation
pkg/federation/apis/federation/install
pkg/federation/apis/federation/v1beta1
pkg/kubelet/qos
pkg/kubelet/types
pkg/master/ports
pkg/util/config
pkg/util/diff
```
These are recursively imported through tests: `rest/client_test.go` imports `pkg/api/testapi`, which imports all the apis and lead to the kubelet imports etc. It seems {{.Deps}} doesn't count the imports caused by test files.

caesarxuchao commented:
@lavalamp all comments address. PTAL. Thanks!

lavalamp commented:
Here's the current structure now.

```
client/
├── discovery
│   └── fake
├── dynamic
├── k8sclient
│   ├── fake
│   └── typed
│       ├── autoscaling
│       ├── batch
│       ├── core
│       └── extensions
├── pkg
│   ├── api
│   │   ├── endpoints
│   │   ├── errors
│   │   ├── install
│   │   ├── meta
│   │   ├── pod
│   │   ├── resource
│   │   ├── service
│   │   ├── testapi
│   │   ├── unversioned
│   │   ├── util
│   │   ├── v1
│   │   └── validation
│   ├── apimachinery
│   │   └── registered
│   ├── apis
│   │   ├── apps
│   │   ├── autoscaling
│   │   ├── batch
│   │   ├── certificates
│   │   ├── componentconfig
│   │   ├── extensions
│   │   ├── policy
│   │   └── rbac
│   ├── auth
│   │   └── user
│   ├── capabilities
│   ├── conversion
│   │   └── queryparams
│   ├── federation
│   │   └── apis
│   ├── fields
│   ├── kubelet
│   │   ├── qos
│   │   ├── server
│   │   └── types
│   ├── labels
│   ├── master
│   │   └── ports
│   ├── runtime
│   │   └── serializer
│   ├── third_party
│   │   └── forked
│   ├── types
│   ├── util
│   │   ├── config
│   │   ├── crypto
│   │   ├── diff
│   │   ├── errors
│   │   ├── flowcontrol
│   │   ├── framer
│   │   ├── hash
│   │   ├── homedir
│   │   ├── httpstream
│   │   ├── integer
│   │   ├── intstr
│   │   ├── json
│   │   ├── labels
│   │   ├── net
│   │   ├── parsers
│   │   ├── rand
│   │   ├── runtime
│   │   ├── sets
│   │   ├── strategicpatch
│   │   ├── testing
│   │   ├── validation
│   │   ├── wait
│   │   └── yaml
│   ├── version
│   └── watch
│       └── versioned
├── rest
├── testing
├── tools
│   ├── auth
│   ├── cache
│   ├── clientcmd
│   │   └── api
│   ├── metrics
│   ├── portforward
│   └── record
└── transport
```

lavalamp commented:
I still think "k8sclient" should be just "kubernetes"; I don't see what sort of collisions it will cause?

lavalamp commented:
We definitely need to do this same process to the util directory...

smarterclayton commented:
I would prefer "kubernetes" as well.

caesarxuchao commented:
Moved clientset to "kubernetes" as suggested.

caesarxuchao commented:
Here's the current structure:
```
.
├── discovery
│   └── fake
├── dynamic
├── kubernetes
│   ├── fake
│   └── typed
│       ├── autoscaling
│       ├── batch
│       ├── core
│       └── extensions
├── pkg
│   ├── api
│   │   ├── endpoints
│   │   ├── errors
│   │   ├── install
│   │   ├── meta
│   │   ├── pod
│   │   ├── resource
│   │   ├── service
│   │   ├── testapi
│   │   ├── unversioned
│   │   ├── util
│   │   ├── v1
│   │   └── validation
│   ├── apimachinery
│   │   └── registered
│   ├── apis
│   │   ├── apps
│   │   ├── autoscaling
│   │   ├── batch
│   │   ├── certificates
│   │   ├── componentconfig
│   │   ├── extensions
│   │   ├── policy
│   │   └── rbac
│   ├── auth
│   │   └── user
│   ├── capabilities
│   ├── conversion
│   │   └── queryparams
│   ├── federation
│   │   └── apis
│   ├── fields
│   ├── kubelet
│   │   ├── qos
│   │   ├── server
│   │   └── types
│   ├── labels
│   ├── master
│   │   └── ports
│   ├── runtime
│   │   └── serializer
│   ├── third_party
│   │   └── forked
│   ├── types
│   ├── util
│   │   ├── config
│   │   ├── crypto
│   │   ├── diff
│   │   ├── errors
│   │   ├── flowcontrol
│   │   ├── framer
│   │   ├── hash
│   │   ├── homedir
│   │   ├── httpstream
│   │   ├── integer
│   │   ├── intstr
│   │   ├── json
│   │   ├── labels
│   │   ├── net
│   │   ├── parsers
│   │   ├── rand
│   │   ├── runtime
│   │   ├── sets
│   │   ├── strategicpatch
│   │   ├── testing
│   │   ├── validation
│   │   ├── wait
│   │   └── yaml
│   ├── version
│   └── watch
│       └── versioned
├── rest
├── testing
├── tools
│   ├── auth
│   ├── cache
│   ├── clientcmd
│   │   └── api
│   ├── metrics
│   ├── portforward
│   └── record
└── transport
```

caesarxuchao commented:
@thockin @smarterclayton could one of you sign this off as well as @lavalamp? Thanks.

smarterclayton commented:
Changes LGTM as a first step (since a fair amount of this can be refined over time).

krousey commented on _staging/src/k8s.io/client-go/copy.sh:
```diff
@@ -0,0 +1,153 @@
+# rewriting package name to v1
+sed -i 's/package api/package v1/g' ${CLIENT_REPO}/pkg/api/v1/ref.go
+# ref.go refers api.Scheme, so manually import /pkg/api
+sed -i "s,import (,import (\n\"${CLIENT_REPO_FROM_SRC}/pkg/api\",g" ${CLIENT_REPO}/pkg/api/v1/ref.go
```
would `goimports` work?

caesarxuchao replied:
I was under the impression we should not rely on `goimports` in our tool. A quick grep shows `go-to-protobuf` uses goimports but that tool is containerized.

krousey commented on _staging/src/k8s.io/client-go/copy.sh:
```diff
@@ -0,0 +1,153 @@
+
+# working in the ${CLIENT_REPO_TEMP} so 'godep save' won't complain about dirty working tree.
+echo "creating the _tmp directory"
+mkdir -p ${CLIENT_REPO_TEMP}
```
Should quote this here and like 30+ other places.

caesarxuchao replied:
Ah, yes. I forgot the best practices you showed me once :)

krousey commented on _staging/src/k8s.io/client-go/copy.sh:
```diff
@@ -0,0 +1,153 @@
+
+echo "move to the client repo"
+# clean the ${CLIENT_REPO}
+ls ${CLIENT_REPO} | grep -v 'copy.sh\|_tmp' | xargs rm -r
```
Does this really need `ls` piped into `grep`? Would `rm -rf copy.sh _tmp` work?

caesarxuchao replied:
It's a `grep -v`. I don't see any equivalent `rm` option.

krousey replied:
Ah. Missed the -v option. Sorry.

If you're curious, the only way to accomplish the negative matching with just `rm` is to `shopt -s extglob`. Then you could `rm -r !(copy.sh|_tmp)`

caesarxuchao replied:
Good to know :)

krousey commented on _staging/src/k8s.io/client-go/copy.sh:
```diff
@@ -0,0 +1,153 @@
+rm -r ${CLIENT_REPO}/vendor
+# remove the pkg/util/net/sets/README.md to silent hack/verify-munge-docs.sh
+# TODO: probably we should convert the README.md a doc.go
+find ./ -name "README.md" | xargs rm
```
`find ./ -name "README.md" -delete`
or
`find ./ -name "README.md" -exec rm {} +`

krousey commented on _staging/src/k8s.io/client-go/copy.sh:
```diff
@@ -0,0 +1,153 @@
+# rewrite pkg/client/record to v1
+gofmt -w -r 'api.a -> v1.a' ${CLIENT_REPO}/pkg/client/record
+# need to call sed to rewrite the strings in test cases...
+find ${CLIENT_REPO}/pkg/client/record -type f -name "*.go" | xargs sed -i "s/api.ObjectReference/v1.ObjectReference/g"
```
add `-print0` to the find command to allow for non-alphanumeric filenames.

caesarxuchao replied:
I added `-print0` to `find` everywhere it's followed by a pipe.

krousey commented on _staging/src/k8s.io/client-go/copy.sh:
```diff
@@ -0,0 +1,153 @@
+# need to call sed to rewrite the strings in test cases...
+find ${CLIENT_REPO}/pkg/client/record -type f -name "*.go" | xargs sed -i "s/api.ObjectReference/v1.ObjectReference/g"
+# rewrite the imports
+find ${CLIENT_REPO}/pkg/client/record -type f -name "*.go" | xargs sed -i 's,pkg/api",pkg/api/v1",g'
```
same here

krousey commented on _staging/src/k8s.io/client-go/copy.sh:
```diff
@@ -0,0 +1,153 @@
+mvfolder pkg/client/unversioned/portforward tools/portforward
+mvfolder pkg/client/metrics tools/metrics
+mvfolder pkg/client/testing/core testing
+if [ -z '$(find ${CLIENT_REPO}/pkg/client -type f -name "*.go")' ]; then
```
I don't think this works. Expressions don't expand in single quotes.

caesarxuchao replied:
changed to:
``` bash
if [ "$(find ${CLIENT_REPO}/pkg/client -type f -name "*.go")" ]; then
```

krousey commented on test/e2e/framework/framework.go:
```diff
@@ -130,6 +135,41 @@ func NewFramework(baseName string, options FrameworkOptions, client *client.Clie
+		sf := sv.Field(i).Interface()
+		data, err := json.Marshal(sf)
+		if err != nil {
+			fmt.Println("CHAO: 1")
```
Remove these debugging statements.

caesarxuchao commented:
In today's sig apimachinary meeting, we voted on the following three alternative client repo organization and the second one wins the popular vote:

* Multiple repos:
  * k8s.io/client-go-1.3/kubernetes
  * k8s.io/client-go-1.4/kubernetes
  * k8s.io/client-go-1.5/kubernetes 
  * …

(the latest one tracks head until we cut that release, then it tracks the release branch in the main repo)

* Single repo, multiple folders: **This one wins the vote in the meeting.**
  * k8s.io/client-go/1.3/kubernetes
  * k8s.io/client-go/1.4/kubernetes
  * k8s.io/client-go/1.5/kubernetes

(same rules as the first one, but lots of code duplication in the repo)

* Single repo, multiple branches
  * We’ve pretty much ruled this out due to it being impossible to `go get`.

caesarxuchao commented:
@k8s-bot gke smoke e2e test, issue #IGNORE
It seems to be a cluster bring up issue.

caesarxuchao commented:
I've moved the contents to k8s.io/client-go/1.4/. I've also addressed @krousey's comments on the copy.sh. @krousey could take another look at the script? Thanks.

krousey commented on _staging/src/k8s.io/client-go/copy.sh:
```diff
@@ -29,6 +29,7 @@ CLIENT_REPO_TEMP="${CLIENT_REPO}"/_tmp
 # working in the ${CLIENT_REPO_TEMP} so 'godep save' won't complain about dirty working tree.
 echo "creating the _tmp directory"
 mkdir -p "${CLIENT_REPO_TEMP}"
+cd ${CLIENT_REPO}
```
Quote.

caesarxuchao replied:
Done. Thanks.

smarterclayton replied:
All other changes are still LGTM (beyond bash which I refuse to be
responsible for understanding)

On Thu, Aug 4, 2016 at 1:32 PM, Chao Xu <notifications@github.com> wrote:

> In _staging/src/k8s.io/client-go/copy.sh
> <https://github.com/kubernetes/kubernetes/pull/29147#discussion_r73566895>
> :
>
> > @@ -29,6 +29,7 @@ CLIENT_REPO_TEMP="${CLIENT_REPO}"/_tmp
> >  # working in the ${CLIENT_REPO_TEMP} so 'godep save' won't complain about dirty working tree.
> >  echo "creating the _tmp directory"
> >  mkdir -p "${CLIENT_REPO_TEMP}"
> > +cd ${CLIENT_REPO}
>
> Done. Thanks.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/29147/files/e7fccc2d9d0bd06a6e5ff96893221053b3457866..a58fce35f3ba60d9a24668b6e5f05422aa27b4d5#r73566895>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p3JtRFkGhiRTKNZWofI44y2T_ni2ks5qciI_gaJpZM4JPNRl>
> .
>

caesarxuchao replied:
Thanks @smarterclayton.

thockin commented:
I really want to look at this but it is 21 commits and 562 files.  Can you
squash all of the non-deps stuff into 1 or 2 commits, and leave the deps in
another one?

On Thu, Aug 4, 2016 at 11:09 AM, Kubernetes Bot <notifications@github.com>
wrote:

> GCE e2e build/test *passed* for commit 8cc5cc8
> <https://github.com/kubernetes/kubernetes/commit/8cc5cc863513d788d781706fe28e4e9ea35dc86a>
> .
>
>    - Test Results
>    <https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/29147/kubernetes-pull-build-test-e2e-gce/51907>
>    - Build Log
>    <http://pr-test.k8s.io/29147/kubernetes-pull-build-test-e2e-gce/51907/build-log.txt>
>    - Test Artifacts
>    <https://console.developers.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/29147/kubernetes-pull-build-test-e2e-gce/51907/artifacts/>
>    - Internal Jenkins Results
>    <http://goto.google.com/prkubekins/job/kubernetes-pull-build-test-e2e-gce//51907>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/29147#issuecomment-237635301>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVIvYaOp4K7VV5lqd5NeQg4lzVUPkks5qcirFgaJpZM4JPNRl>
> .
>

thockin commented:
Why "_staging"?  Do we want this excluded from tools?  I ask because this
PR takes what were (relatively) simple `find` commands which exclude `_*`
and changes them to include `_staging`.  at that point, why not call it
`staging`?

Does the whole "check in a symlink in vendor/ thing actually work, and pass
verify-godeps?

On Thu, Aug 4, 2016 at 10:51 PM, Tim Hockin <notifications@github.com>
wrote:

> I really want to look at this but it is 21 commits and 562 files. Can you
> squash all of the non-deps stuff into 1 or 2 commits, and leave the deps in
> another one?
>
> On Thu, Aug 4, 2016 at 11:09 AM, Kubernetes Bot <notifications@github.com>
> wrote:
>
> > GCE e2e build/test *passed* for commit 8cc5cc8
> > <https://github.com/kubernetes/kubernetes/commit/
> 8cc5cc863513d788d781706fe28e4e9ea35dc86a>
> > .
> >
> > - Test Results
> > <https://k8s-gubernator.appspot.com/build/kubernetes-
> jenkins/pr-logs/pull/29147/kubernetes-pull-build-test-e2e-gce/51907>
> > - Build Log
> > <http://pr-test.k8s.io/29147/kubernetes-pull-build-test-
> e2e-gce/51907/build-log.txt>
> > - Test Artifacts
> > <https://console.developers.google.com/storage/browser/
> kubernetes-jenkins/pr-logs/pull/29147/kubernetes-pull-
> build-test-e2e-gce/51907/artifacts/>
> > - Internal Jenkins Results
> > <http://goto.google.com/prkubekins/job/kubernetes-
> pull-build-test-e2e-gce//51907>
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/kubernetes/kubernetes/pull/
> 29147#issuecomment-237635301>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/
> AFVgVIvYaOp4K7VV5lqd5NeQg4lzVUPkks5qcirFgaJpZM4JPNRl>
> > .
> >
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/29147#issuecomment-237761835>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVFpnoQ5_5dJHfb6Q-rwkMDZpm-0iks5qcs9ogaJpZM4JPNRl>
> .
>

caesarxuchao commented:
@k8s-bot unit test this, issue #IGNORE

caesarxuchao commented:
@thockin i'll re-arrange the commits once all tests pass again.

Regarding "_staging", I remember I had a good reason to add an underscore, but perhaps that's before your makefile changes went in and is not necessary now. I'll try.

Regarding godeps, I verified it all works, but I should try it again now.

caesarxuchao commented:
For the record, proto will complain `proto: duplicate proto type registered ...` if the main repo client and the client repo client are imported together. Currently the e2e tests import both.

caesarxuchao commented:
@thockin, I checked this godep workflow and it works: remove Godeps/, remove everything including the symlink from vendor/, then run hack/godep-save.sh successfully restored the repo's content.

I'll try rename to `staging`.

caesarxuchao commented:
I changed the names in `proto.RegisterType` from `k8s.io.kubernetes.pkg...` to `k8s.io.client-go.1.4.pkg...` and that solves the conflicts. @smarterclayton said proto import mechanism doesn't handle "." in package name well, so the name "1.4" might cause trouble later on.

Commit:
move pkg/util/ files to their own package to reduce client-go dependencies

Commit:
added copy.sh which cuts the client repo; made necessary changes to scripts in
the repo to cope with the client repo staging area; added symlink
vendor/k8s.io/client-go which pointed to the staging area.

Commit:
added a e2e test that uses the staged client; made necessary changes to gcp auth plugin to pass the gke tests

caesarxuchao commented:
@k8s-bot node e2e test this, issue #IGNORE

thockin commented on hack/godep-save.sh:
```diff
@@ -31,4 +34,7 @@ REQUIRED_BINS=(
 pushd "${KUBE_ROOT}" > /dev/null
   GO15VENDOREXPERIMENT=1 ${GODEP} save "${REQUIRED_BINS[@]}"
+  if [ ! -e "vendor/k8s.io/client-go" ]; then
+    ln -s ../../staging/src/k8s.io/client-go vendor/k8s.io/client-go
```
Can you comment here about this - it is tricky and likely to be misunderstood

caesarxuchao replied:
Will do.

caesarxuchao replied:
Done.

Commit:
generated staging area

caesarxuchao commented:
@thockin it's ready for review. Only the "added copy.sh..." and the "added an e2e test..." commits are interesting. And the latest client repo structure is:

```
.
└── 1.4
    ├── discovery
    │   └── fake
    ├── dynamic
    ├── kubernetes
    │   ├── fake
    │   └── typed
    │       ├── authorization
    │       ├── autoscaling
    │       ├── batch
    │       ├── core
    │       └── extensions
    ├── pkg
    │   ├── api
    │   │   ├── endpoints
    │   │   ├── errors
    │   │   ├── install
    │   │   ├── meta
    │   │   ├── pod
    │   │   ├── resource
    │   │   ├── service
    │   │   ├── testapi
    │   │   ├── unversioned
    │   │   ├── util
    │   │   ├── v1
    │   │   └── validation
    │   ├── apimachinery
    │   │   └── registered
    │   ├── apis
    │   │   ├── apps
    │   │   ├── authentication
    │   │   ├── authorization
    │   │   ├── autoscaling
    │   │   ├── batch
    │   │   ├── certificates
    │   │   ├── componentconfig
    │   │   ├── extensions
    │   │   ├── policy
    │   │   └── rbac
    │   ├── auth
    │   │   └── user
    │   ├── capabilities
    │   ├── conversion
    │   │   └── queryparams
    │   ├── federation
    │   │   └── apis
    │   ├── fields
    │   ├── kubelet
    │   │   ├── qos
    │   │   ├── server
    │   │   └── types
    │   ├── labels
    │   ├── master
    │   │   └── ports
    │   ├── runtime
    │   │   └── serializer
    │   ├── third_party
    │   │   └── forked
    │   ├── types
    │   ├── util
    │   │   ├── clock
    │   │   ├── config
    │   │   ├── crypto
    │   │   ├── diff
    │   │   ├── errors
    │   │   ├── flowcontrol
    │   │   ├── framer
    │   │   ├── hash
    │   │   ├── homedir
    │   │   ├── httpstream
    │   │   ├── integer
    │   │   ├── intstr
    │   │   ├── json
    │   │   ├── labels
    │   │   ├── net
    │   │   ├── parsers
    │   │   ├── rand
    │   │   ├── runtime
    │   │   ├── sets
    │   │   ├── strategicpatch
    │   │   ├── testing
    │   │   ├── uuid
    │   │   ├── validation
    │   │   ├── wait
    │   │   └── yaml
    │   ├── version
    │   └── watch
    │       └── versioned
    ├── rest
    ├── testing
    ├── tools
    │   ├── auth
    │   ├── cache
    │   ├── clientcmd
    │   │   └── api
    │   ├── metrics
    │   ├── portforward
    │   └── record
    └── transport
```

thockin commented:
FWIW I do not intend to properly review this unless there's something
unique I can offer.  Otherwise the GOPATH stuff LGTM

On Sat, Aug 6, 2016 at 4:08 PM, Kubernetes Bot <notifications@github.com>
wrote:

> GCE e2e build/test *passed* for commit 1507e5e
> <https://github.com/kubernetes/kubernetes/commit/1507e5ec0ba27db7f3782666897c3b8a2f054995>
> .
>
>    - Test Results
>    <https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/29147/kubernetes-pull-build-test-e2e-gce/52244>
>    - Build Log
>    <http://pr-test.k8s.io/29147/kubernetes-pull-build-test-e2e-gce/52244/build-log.txt>
>    - Test Artifacts
>    <https://console.developers.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/29147/kubernetes-pull-build-test-e2e-gce/52244/artifacts/>
>    - Internal Jenkins Results
>    <http://goto.google.com/prkubekins/job/kubernetes-pull-build-test-e2e-gce//52244>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/29147#issuecomment-238053668>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVNhHb1BTHJfp3cHcJTvYEvFHGqEaks5qdRQFgaJpZM4JPNRl>
> .
>

smarterclayton commented:
This is now LGTM, I'd like to merge it unless there are any other comments about the bash scripts.

caesarxuchao commented:
@k8s-bot test this, issue #IGNORE

The jenkins verfication test failed without any log, I don't know how to trigger only the verification test.

caesarxuchao commented:
Thanks for the reviews. @lavalamp approve the PR IRL. I'll apply the label.

seh commented:
I understand that this PR focused on _staging_ the migration to a separate repository. Is there a related issue that tracks the actual creation of that separate repository? Searching today, I'm not able to find such a repository within the "kubernetes" organization.

caesarxuchao commented:
@seh, we are working on creating the separate repository. I'll update #28559 to report the progress.
